### PR TITLE
UDP Server and Tracelet API

### PR DESCRIPTION
**Pull request by FranziskaKnoedtel**
Added UDP Socket implementation.
Added FrameHandshake implementation.
Added Tracelet Server API.
Added examples to use Tracelet Server API.
Changed channels ReadMsg() implementation (fix timeout handling that blocking read does not continue on timeout)

### CODE REVIEW

File: server/udp_server.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/ci4rail/io4edge-client-go/transport"
+	"github.com/ci4rail/io4edge-client-go/transport/socket"
+)
+
+// UDPServer represents a server for io4edge devices
+type UDPServer struct {
+	lis *socket.UDPListener
+}
+
+// NewServer creates a new function client from a socket with the specified address.
```
Comment (KlausPopp) [inline server/udp_server.go:35]: rename to NewUDPServer to distinguish from future TCP Server

File: tracelet/stream_server.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+// Channel represents a connection to a tracelet
+type Channel struct {
+	ch      *client.Channel
+	timeout time.Duration
+}
+
+// NewTraceletServer creates a new tracelet server
+func NewTraceletServer(port string, timeout time.Duration) *Server {
```
Comment (KlausPopp) [inline tracelet/stream_server.go:29]: rename to NewTraceletUDPServer

why is port a string?

File: tracelet/stream_server.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	log "github.com/sirupsen/logrus"
+
+	"github.com/ci4rail/io4edge-client-go/client"
+	"github.com/ci4rail/io4edge-client-go/server"
+	"github.com/ci4rail/io4edge-client-go/transport"
+	"github.com/ci4rail/io4edge_api/tracelet/go/tracelet"
+)
+
+// Server represents a tracelet server
+type Server struct {
```
Comment (KlausPopp) [inline tracelet/stream_server.go:16]: rename to UDPServer

File: tracelet/stream_server.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+				return
+			} else if err != nil {
+				continue
+			}
+			// go through all subscriptions and search for matches
+			for pattern, stream := range *subscriptions {
+				// check if pattern matches tracelet id
+				match, err := regexp.MatchString(pattern, msg.TraceletId)
+				if err != nil {
+					log.Errorf("Error matching pattern: %v", err)
```
Comment (KlausPopp) [inline tracelet/stream_server.go:113]: This is no error. It only means that noone has subscribed

File: tracelet/stream_server.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+					stream <- msg
+				}
+			}
+		}
+	}()
+}
+
+// Close closes the tracelet channel
+func (t *Channel) Close() {
+	log.Printf("Closing connection")
```
Comment (KlausPopp) [inline tracelet/stream_server.go:124]: Change to Debug

File: tracelet/stream_server.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+func (s *Server) Unsubscribe(traceletPattern string) {
+	delete(s.subscriptions, traceletPattern)
+}
+
+// ReadData reads a single message from the tracelet channel
+func (t *Channel) readData() (*tracelet.TraceletToServer, error) {
+	msg := &tracelet.TraceletToServer{}
+	err := t.ch.ReadMessage(msg, t.timeout)
+	if err == transport.ErrTimeout {
+		log.Printf("Read timeout -> close socket")
```
Comment (KlausPopp) [inline tracelet/stream_server.go:87]: Change to debug

File: transport/socket/udp_socket.go [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+import (
+	"net"
+	"time"
+
+	"github.com/ci4rail/io4edge-client-go/transport"
+	log "github.com/sirupsen/logrus"
+)
+
+// ReadBufferSize is the size of the buffer for incoming messages of one connection
+var ReadBufferSize = 10
```
Comment (KlausPopp) [inline transport/socket/udp_socket.go:28]: Is this intentionally exported? Better use const

File: transport/socket/udp_socket.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+			log.Debugf("Received on new socket for remote %s", remoteAddr.String())
+			// create new socket
+			s := &UDPSocket{
+				conn:         c,
+				remoteAddr:   remoteAddr,
+				readDeadline: time.Time{},
+				readData:     make(chan []byte, ReadBufferSize),
+			}
+			go func() {
+				select {
```
Comment (KlausPopp) [inline transport/socket/udp_socket.go:176]: DRY: factor out duplicate code

**Review by KlausPopp — CHANGES_REQUESTED** [review]

File: transport/frame.go [hunk 9]
```diff
@@ -89,22 +91,41 @@ func (fs *FramedStream) writeBytesSafe(payload []byte) error {
...
+		netErr, ok := err.(net.Error)
+		if ok && netErr.Timeout() {
+			err = ErrTimeout
+		}
 		return nil, err
 	}
 
 	length, err := fs.readLength()
 	if err != nil {
+		netErr, ok := err.(net.Error)
```
[the remark this answers is not in the document]
Comment (FranziskaKnoedtel) [reply transport/frame.go:113]: I did the error handling according to: https://go.dev/blog/error-handling-and-go#the-error-type
There is an example for how to handle errors of the net package.

File: examples/tracelet/specific_tracelet/main.go [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+		loc := msg.GetLocation()
+		log.Printf("Received tracelet 3 location: %v", loc.Gnss)
+	}
+}
+
+func main() {
+	log.SetLevel(log.InfoLevel)
+	server := tracelet.NewTraceletServer("11002", time.Second*5)
+
+	go catchTracelet1Stream(server.Subscribe("TRACELET-1"))
```
[the remark this answers is not in the document]
Comment (FranziskaKnoedtel) [reply examples/tracelet/specific_tracelet/main.go:40]: The regex bounderies are added now internally in Subscribe() to the pattern.

**Review by KlausPopp — APPROVED** [review]

**Pull request merged**